### ISSUE TITLE
perf(rl): 细分并优化 PPO rollout 的 GPU planner 热路径

### ISSUE THREAD

**Issue by xcz0** [xcz0/Eco-AutoDrive#74]
## Goal

定位并降低 PPO policy-guided rollout 中 GPU planner 热路径的耗时，同时保持当前 sampler、guidance、随机流和训练语义不变。

## Context

2026-08-28 在 commit `d54553bcbe8294a10596bfbe5666c3ba57129b7a`、`rtx3050_laptop` resource profile 上完成了真实 CUDA BF16 smoke training：

- 4 PPO updates，128 transitions，训练成功；
- policy hash 发生变化，frozen planner hash 保持不变；
- profile 配置上限为 4 rollout workers，但两个逻辑 scenarios 使实际 worker pool 为 2 个进程，每个进程 8 Torch threads。

与实际 rollout 形状一致的 `B=2 × 16 transitions` 热态 benchmark（PPO 单 epoch）测得：

- end-to-end：3.800 s/update；
- planner decision + bootstrap：3.536 s/update，占 93.0%；
- MetaDrive vector environment：0.084 s/update，占 2.2%；
- collection overhead：0.093 s/update，占 2.5%；
- throughput：8.59 transitions/s；
- cold start：11.92 s，是热态的 3.1 倍。

扩展性测量的吞吐为 B=1/2/4/8 对应 4.72/8.29/15.54/28.88 transitions/s。进程内 Torch threads=1/2/4/8 对 B=4 吞吐没有显示出稳定差异，因此当前证据不支持优先优化 MetaDrive、共享内存或线程预算。

当前 rollout benchmark 无法用 `benchmark.ppo_epochs=4` 覆盖真实训练：它把 scheduler horizon 设置为 update 数，第二个 optimizer step 会触发 `PPO update would exceed the configured scheduler horizon`。这阻塞了对真实 PPO 配置的完整分阶段测量。

## Status update — 2026-08-28

计时与 benchmark 契约修复已在当前工作树完成并通过验证，尚未提交；本轮没有落地性能优化，Issue 保持开启。

- rollout benchmark 现在显式使用 `ppo_minibatch_size`，scheduler horizon 按 `updates × epochs × (sample_count / minibatch_size)` 计算。B=2×16、4 epochs 时每个 update 为 8 个 optimizer steps；B=4 时为 16 个。
- profiler 分离 host-call wall、CUDA Event accelerator duration、collate、deferred audit transfer 与 audit resolve wait。bootstrap 的异步 CUDA 工作在 profiling 模式下由自身 Event 归因；profiling 关闭时不创建 Event、不增加同步。
- 正式 RTX 3050 测量使用固定 seeds、B=2/B=4、1 warmup + 3 measured updates、3 repeats，结果登记为 E-026：
  - B=2：collection 4.718 s/update，PPO 0.355 s/update，end-to-end 5.023 s/update，6.78 transitions/s；
  - B=4：collection 4.721 s/update，PPO 0.597 s/update，end-to-end 5.313 s/update，13.56 transitions/s。
- 计时证据显示两次 DDIM 路径仍主导 planner wall；CUDA/host wall、跨 stream audit transfer 与 worker busy 是不同视角，不可相加为总墙钟。
- E-026 记录路径：`docs/experiments/records/e-026-issue74-corrected-rollout-profiler.md`。旧 artifact 仅保留 total wall/throughput 的历史参考；旧 `planner_decision`、`planner_bootstrap` 和 `collection_overhead` 分项因异步归因及边界错误视为无效。
- 验证：CPU 49 passed；GPU 1 passed；simulator 5 passed；Ruff 与 Pyright 通过。

剩余工作是基于修正后的 profiler 选择语义保持的最小优化，并在相同 workload/seeds 下完成至少 3 次 B=2/B=4 before/after 对照。当前没有证据支持修改 resource profile 默认值。

## Scope

- 修正 rollout benchmark 的 PPO optimizer-step/scheduler horizon 计算，使配置的 epochs 和 minibatch 语义能够覆盖真实训练形状。
- 在现有显式 profiling hook 下细分 planner wall time，至少区分：
  - reference DDIM pass / `prepare_policy_guidance`；
  - Exploration Policy forward 与 action sampling；
  - guided DDIM pass / `complete_policy_guidance`；
  - execution trajectory D2H、deferred audit resolve 及显式 CUDA/host 同步。
- 用固定 seed 对 B=2 与 B=4 做可重复测量，确认小 batch、两次 DDIM pass和同步点各自的贡献。
- 仅在 profiling 证据明确时实现最小的、语义保持的优化，并给出 before/after 数据。
- 若 benchmark 的持久执行契约发生变化，同步更新 `docs/agents/system-contract.md`。

## Non-goals

- 不改变 DDIM step 数、guidance 数学、reward、GAE/PPO 定义、precision 或确定性设置。
- 不把增加 scenarios/改变 PPO batch 当作透明运行时优化。
- 不引入多 GPU、TorchRL Collector、fallback sampler 或近似推理路径。
- 不以选择更有利的 seed 或跳过 cold-start 数据制造性能提升。

## Acceptance criteria

- [x] `B=2 × 16 transitions`、`ppo_epochs=4` 的 rollout benchmark 能完成全部 warmup/measured updates，scheduler horizon 覆盖实际 optimizer steps。
- [x] benchmark JSON 保存上述 planner 子阶段的原始 samples、median/minimum/maximum，并保持 profiling 关闭时不增加 CUDA 同步。
- [x] 在相同 seed、sampler、guidance、precision 和 workload 下至少完成 3 次重复的 B=2/B=4 before/after 对照。
- [x] 结论明确区分 algorithmic diffusion cost、CUDA/host synchronization、MetaDrive worker time 与一次性 cold start。
- [x] 若落地优化，相关训练/rollout/simulator 测试通过，policy 可更新且 frozen planner 不变；若没有安全优化，则以 profiling 证据明确记录瓶颈边界，不提交改变研究语义的替代方案。
- [x] 不修改 resource profile 默认值，除非独立测量证明该 host scheduling 变更稳定有效。

**Comment by xcz0** [comment]
我看了一下当前 `planner decision + bootstrap` 的实现，建议先修 profiling 口径，再做执行路径优化。现在有两个比较关键的问题：

1. **当前 decision / bootstrap 计时对 CUDA 不对称，分项结果可能失真**

`VectorRolloutCollector._collect_group()` 用 `perf_counter()` 包住 `decide_batch()` / `bootstrap_value_batch()`。但 `decide_batch()` 最后 `copy_execution_trajectory()` 会显式 `cuda.current_stream().synchronize()`，所以 decision 基本包含真实 GPU wall time；而 `bootstrap_value_batch()` 返回 GPU tensor，没有同步，`perf_counter()` 很可能只量到 kernel launch。未完成的 bootstrap work 会在后续同步点被等待，从而出现：

- `planner_bootstrap_wall_s` 偏小
- `planner_decision_wall_s` 偏大

另外 `planner_started` 放在 `collate_observations(...)` 之前，因此 `planner_decision_wall_s` 还包含 CPU collate 时间。

建议 benchmark 模式改用 CUDA Events，并单独拆：`collate / H2D / encoder / reference DDIM / policy / guided DDIM / execution D2H / bootstrap`。生产路径不要为了计时额外全局 synchronize。

2. **bootstrap 的结构性成本非常高：为了一个 scalar V(s)，又完整跑了一遍 reference planner**

当前 `bootstrap_value_batch()`：

```python
moved = self._fabric.to_device(raw_observation)
noise = _slot_noise(...)
prepared = self._planner.prepare_policy_guidance(moved, noise, diffusion_generators)
outputs = self._policy.forward_tensordict(...)
value = outputs["state_value"]
```

而 `prepare_policy_guidance()` 本身包含：

```text
observation normalize
→ scene encoder + route encoder
→ reference DDIM-5
```

所以 bootstrap 并不是轻量 critic forward，而是接近一次 unguided planner inference，只是为了拿 `state_value`。

如果允许调整 RL architecture，最值得做的是 **asymmetric actor/critic**：

```text
Actor:  scene + navigation + diffusion reference → guidance action
Critic: scene + navigation (+ ego state)         → V(s)
```

让 critic 不依赖 `reference_trajectory`，bootstrap 就可以从：

```text
encoder + route encoder + 5×denoiser + policy trunk
```

降到：

```text
encoder + route encoder + lightweight value head
```

这应该是 bootstrap 最大的潜在收益。

3. **decision 本身一次实际上跑了两遍 DDIM-5**

`prepare_policy_guidance()` 先跑 reference DDIM-5；policy 选 action 后，`complete_policy_guidance()` 再跑 guided DDIM-5。guided pass 每个 step 还会 `requires_grad_(True)` + `torch.autograd.grad()`，所以一次 decision 大致是：

```text
encoder
+ 5 × denoiser forward
+ policy
+ 5 × denoiser forward + input-gradient backward
```

这部分是 decision 的主要 FLOPs，微优化 TensorDict 不会改变数量级。

4. **hot path 里有两个不必要的同步/检查，建议先删**

`_run_guided()`：

```python
if torch.count_nonzero(action).item() == 0:
```

`.item()` 会触发 CUDA host-device synchronization。对于连续 Beta policy，整批 action 精确全 0 基本不会发生。建议 learned-policy path 直接走 guided sampler，把 zero-action fast path 留给 fixed/reference guidance。

另外：

```python
torch.testing.assert_close(result.guidance_action, action.guidance_action)
```

也不建议放在每个 planning cycle 的 production hot path；按当前实现它基本是在检查刚传进去的 tensor。可以放 debug/test flag。

5. **如果暂时不改 critic，bootstrap 可以延迟并合批**

现在某个 slot 一旦 truncated / rollout_limit，就立即对这些 tail slots 调 `bootstrap_value_batch()`。如果 physical capacity=8，但本轮只有 1～2 个 tail，GPU batch utilization 很差。

bootstrap 不影响下一次 env reset/decision，可以暂存：

```text
pending_bootstrap = [(episode, next_observation, rng_state), ...]
```

到 pending queue 满，或者本次 `collect()` 结束，再统一做 `bootstrap_value_batch()`。这不减少 FLOPs，但会提高小 batch bootstrap 的 GPU utilization。当前 benchmark 已经有 `bootstrap_batch_fill_ratio`，可以直接验证收益。

6. **低风险 memory / launch 优化**

`build_training_decision()` 目前对所有 context 做：

```python
value.detach().clone()
```

scene tokens / reference trajectory 都会发生额外 D2D copy + allocation。只要确认生命周期内不会原地修改，可以改成仅 `detach()`，只对确实需要 ownership isolation 的字段 clone。

另外 `_slot_noise()` 是 per-slot `randn` 后 `torch.cat`，Beta action sampling 也是 Python loop + per-slot kernel；batch 1/2/4/8 时如果 profiler 里 launch overhead 明显，可以预分配 tensor 后按 slice 填充，减少 `cat` / allocation。

7. **DDIM 内部还有一些稳定的小优化**

当前每步 constraint：

```python
constrained = sample.reshape(...).clone()
constrained[:, :, 0] = current_states
```

reference + guided 两遍会重复很多次整 tensor clone。可以检查 scheduler/autograd aliasing 后，对新产生的 sample 做安全 in-place constraint，减少 clone。

`ddim_stochasticity == 0` 时，每 step 目前仍可能 `zeros_like(sample)`；如果 diffusers `eta=0` 路径允许 `variance_noise=None`，也可以避免这些 allocation。

8. **最后再做 `torch.compile`，优先 compile decoder**

不建议第一步 compile 整个 `FabricRolloutRuntime.decide_batch()`，因为里面有 Python object、TensorDict、Generator、diffusers scheduler、host transfer，容易 graph break。

优先尝试：

```python
planner.model.encoder = torch.compile(planner.model.encoder, mode="reduce-overhead")
planner.model.decoder.dit = torch.compile(planner.model.decoder.dit, mode="reduce-overhead")
policy = torch.compile(policy, mode="reduce-overhead")
```

其中 `decoder.dit` 每个 decision 会被调用约 10 次，优先级最高。对 batch 1/2/4/8 分别 warmup 后再看 steady-state。

建议实际落地顺序：

```text
P0  CUDA Event phase profiler + collate 单独计时
P1  去掉 count_nonzero(...).item() / hot-path assert / 多余 clone
P2  compile decoder + 延迟合批 bootstrap
P3  critic 与 diffusion reference 解耦
```

我认为最值得优先验证的两个方向是：

```text
decision:  去同步点 + compile decoder
bootstrap: critic 不再依赖 reference trajectory
```

前者直接减少 hot-path latency，后者可以从结构上去掉 bootstrap 的 5-step diffusion sampling。

**Comment by xcz0** [comment]
已按顺序完成本轮本地实现与验证，结果如下。

### 1. 局部编译 DiT

- 新增必填配置 `training.planner_compile_mode: eager | dit_reduce_overhead`，训练 runtime 与 benchmark 均显式传递并记录实际模式。
- `dit_reduce_overhead` 仅编译 `decoder.dit` 的 bound `forward`：
  `torch.compile(mode="reduce-overhead", fullgraph=True, dynamic=False)`。
- 不替换注册模块，不修改 state-dict key；仅允许 CUDA，任何失败直接报错，不 fallback eager。
- 本机 compile smoke 在首次 Inductor backend 初始化时失败：
  `torch._inductor.exc.TritonMissing: Cannot find a working triton installation`。
  当前 lock 只在 Linux 安装 Triton，因此没有成功 graph、recompile 或稳态 B=2/B=4 样本。
- 编译性能门槛未进入判断，PPO base profile 保持 `eager`。

### 2. 同步清理

- 删除 production `torch.testing.assert_close`，planner guidance action identity 改由单元测试保证。
- fixed/reference zero-action fast path 保持不变。
- learned-policy 的 `count_nonzero(...).item()` 曾改为 device-side selection；完整矩阵未证明收益（B=2 5.293 s，B=4 5.761 s），已全部撤销并保留原检查。

### 3. DDIM allocation

- `ddim_stochasticity=0` 时改为向 diffusers 传递 `variance_noise=None`，不再创建 `zeros_like(sample)`。
- 单元测试确认与显式零 tensor 的输出逐位一致，generator RNG state 逐位不变。

### 4. Benchmark

使用与 E-026 相同的 B=2/B=4、固定 seeds、BF16、DDIM-5、1 warmup、3 measured、3 repeats。

- eager baseline collection：B=2 `4.712 s`，B=4 `5.363 s`。
- final safe cleanup：B=2 `5.246 s`，B=4 `5.690 s`。
- 未修改的 prepare/guided DDIM 阶段同时出现 13%–18% 漂移；随后恢复旧实现的 thermal control 又得到 B=2 `3.754 s`、B=4 `5.372 s`，且 B=4 repeat medians 为 `4.243/7.998/5.372 s`。

因此本机温度/功耗漂移显著大于两个微优化，不把全局 before/after 差值归因于实现，也不主张 throughput 改善。两项 cleanup 仅依据逐值/RNG 契约测试保留。显存峰值未变化，无 OOM、非有限值或 simulator failure。

### 5. 验证

- targeted CPU：20 passed（另复跑严格配置测试 10 passed）
- `just test-gpu`：1 passed
- `just test-sim`：5 passed
- slow 真实 checkpoint 两-transition PPO：1 passed，policy 更新且 frozen planner hash 不变
- Ruff：通过
- Pyright：0 errors / 0 warnings

完整记录已写入 `docs/experiments/records/e-027-issue74-rollout-hotpath-optimization.md`，system contract 与实验索引同步更新。未处理既有重复 E-026 编号，也未实施计划列出的非目标架构改动。
